tests/testsuite/warning_override.rs

```diff
@@ -0,0 +1,100 @@
+//! Tests for overriding warning behavior using `build.warnings` config option.
+
+use std::sync::LazyLock;
+
+use cargo_test_support::{cargo_test, project, str, Project};
+use snapbox::data::Inline;
+
+const ALLOW_CLEAN: LazyLock<Inline> = LazyLock::new(|| {
+    str![[r#"
+[CHECKING] foo v0.0.1 ([ROOT]/foo)
+[WARNING] unused variable: `x`
+...
+[WARNING] `foo` (bin "foo") generated 1 warning
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]]
+});
+
+const ALLOW_CACHED: LazyLock<Inline> = LazyLock::new(|| {
+    str![[r#"
+[WARNING] unused variable: `x`
+...
+[WARNING] `foo` (bin "foo") generated 1 warning
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]]
+});
+
+static WARN: LazyLock<Inline> = LazyLock::new(|| {
+    str![[r#"
+...
+[WARNING] unused variable: `x`
+...
+[WARNING] `foo` (bin "foo") generated 1 warning
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]]
+});
+
+const DENY: LazyLock<Inline> = LazyLock::new(|| {
+    str![[r#"
+...
+[WARNING] unused variable: `x`
+...
+[WARNING] `foo` (bin "foo") generated 1 warning
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]]
+});
+
+fn make_project(main_src: &str) -> Project {
+    project()
+        .file(
+            "Cargo.toml",
+            &format!(
+                r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                edition = "2021"
+            "#
+            ),
+        )
+        .file("src/main.rs", &format!("fn main() {{ {} }}", main_src))
+        .build()
+}
+
+#[cargo_test]
+fn rustc_caching_allow_first() {
+    let p = make_project("let x = 3;");
+    p.cargo("check").with_stderr_data(ALLOW_CLEAN.clone()).run();
```

Comment:
Do we need a clippy test to make sure it works there as well?

CC @Muscraft since this might have come up when talking about the "fixable" summary

Reply:
Added a clippy test